src/engine/IdTable.h

```diff
@@ -13,557 +13,751 @@
+    // rows invovlved in an assigment does not manage it's data, but rather
+    // functions as a view into an IdTable
+    if (_allocated) {
+      // If we manage our own storage recreate that to fit the other row
+      delete[] _data;
       if (other._allocated) {
+        // Both rows manage their own storage so we can take advantage
+        // of move semantics.
+        _data = other._data;
+        _cols = other._cols;
+
+        // otherwise the data will be deleted unexpectedly
         other._data = nullptr;
-      } else {
-        std::memcpy(_data, other._data, sizeof(Id) * _cols);
-      }
-    }
 
-    Row& operator=(const Row& other) {
-      // Check for self assignment.
-      if (&other == this) {
         return *this;
-      }
-      if (_allocated) {
-        // If we manage our own storage recreate that to fit the other row
-        delete[] _data;
+      } else {
         _data = new Id[other._cols];
         _cols = other._cols;
       }
-      // Copy over the data from the other row to this row
-      if (_cols == other._cols && _data != nullptr && other._data != nullptr) {
-        std::memcpy(_data, other._data, sizeof(Id) * _cols);
-      }
-      return *this;
     }
-
-    Row& operator=(Row&& other) {
-      // Check for self assignment.
-      if (&other == this) {
-        return *this;
-      }
-      // This class cannot use move semantics if at least one of the two
-      // rows invovlved in an assigment does not manage it's data, but rather
-      // functions as a view into an IdTable
-      if (_allocated) {
-        // If we manage our own storage recreate that to fit the other row
-        delete[] _data;
-        if (other._allocated) {
-          // Both rows manage their own storage so we can take advantage
-          // of move semantics.
-          _data = other._data;
-          _cols = other._cols;
-
-          // otherwise the data will be deleted unexpectedly
-          other._data = nullptr;
-
-          return *this;
-        } else {
-          _data = new Id[other._cols];
-          _cols = other._cols;
-        }
-      }
-      // Copy over the data from the other row to this row
-      if (_cols == other._cols && _data != nullptr && other._data != nullptr) {
-        std::memcpy(_data, other._data, sizeof(Id) * _cols);
-      }
-      return *this;
+    // Copy over the data from the other row to this row
+    if (_cols == other._cols && _data != nullptr && other._data != nullptr) {
+      std::memcpy(_data, other._data, sizeof(Id) * _cols);
     }
+    return *this;
+  }
 
-    bool operator==(const Row& other) const {
-      bool matches = _cols == other._cols;
-      for (size_t i = 0; matches && i < _cols; i++) {
-        matches &= _data[i] == other._data[i];
-      }
-      return matches;
+  bool operator==(const Row& other) const {
+    bool matches = _cols == other._cols;
+    for (size_t i = 0; matches && i < _cols; i++) {
```

Comment:
Ahh this is one of these cases that make you wonder if a shortcut exit at first difference is faster or slower, depending on how clever the hardware people made branch prediction and pipelining. Maybe we should use `stdd:memcmp` and have the libc deal with that? Fun fact s390x of course has a CPU instruction for that :D https://github.com/torvalds/linux/blob/8f3d9f354286745c751374f5f1fcafee6b3f3136/arch/s390/lib/string.c#L343

Reply:
`memcmp` is not the solution, because it immediately breaks the code as soon as `Ids` are not plain ints and have a custom comparison.

`return _cols == other._cols && std::equal(_data, _data + _cols, other._data);`

(`std::equal` should become `std::memcmp` for builtin types.)
Also this fixes the bug, that in case the two sizes don't match (which should never happen in calling code), there is an out of bounds read.